pkg/apis/kops/validation/instancegroup.go

```diff
@@ -210,6 +214,23 @@ func CrossValidateInstanceGroup(g *kops.InstanceGroup, cluster *kops.Cluster, st
 	return allErrs
 }
 
+func ValidateMasterInstanceGroup(g *kops.InstanceGroup, cluster *kops.Cluster) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for _, etcd := range cluster.Spec.EtcdClusters {
+		hasEtcd := false
+		for _, m := range etcd.Members {
+			if fi.StringValue(m.InstanceGroup) == g.ObjectMeta.Name {
+				hasEtcd = true
+				break
+			}
+		}
+		if !hasEtcd {
+			allErrs = append(allErrs, field.NotFound(field.NewPath("spec", "etcdClusters").Key(etcd.Name), g.ObjectMeta.Name))
```

Comment:
```suggestion
			allErrs = append(allErrs, field.NotFound(field.NewPath("spec", "etcdClusters").Key(k).Key("name"), etcd.Name, "has no member for master instancegroup " + g.ObjectMeta.Name))
```

where k is the index from the range over EtcdClusters

Reply:
Actually, the NotFound error is the other way around. It is the instancegroup's ObjectMeta.Name that wasn't found, so it should be `field.NewPath("objectMeta", "name")`

Reply:
Actually, the NotFound error is the other way around. It is the instancegroup's Name that isn't found in the etcdcluster's members. So it should be `field.NewPath("objectMeta", "name")`

Reply:
Thanks for spotting this.

`NotFound` unfortunately doesn't take a third description argument, so I was looking hard at how to make this more clear.

If I understand the `NotFound` description correctly, the error's field should point to the missing value, not the "cause" of the missing value (which may not even be a path). Since the validation happens as a part of a lookup in the etcdSpec, I found this direction to be correct.

Question is how to make this clear to users. The most programatically correct solution would be `field.NotFound(field.NewPath("spec", "etcdClusters").Index(k).Key("instanceGroup")` but not sure that is the most understandable for users as it won't contain the etcd cluster name.

Reply:
I changed the path to be more correct + added a deep validation test to see how this will look like to the end user.

Reply:
The value that's missing is the instancegroup name.

I think we need to use `field.Forbidden` here. I believe `field.NotFound` is for when there is an obvious list of possible choices for the field's value. `field.NotFound` on a path inside spec.etcdClusters would be appropriate for a validation that checks if an etcdCluster Member's InstanceGroup field value was actually a defined InstanceGroup (possibly restricting to instancegroups with role master).

Reply:
Would it not be reasonable for a five-master-IG cluster to have a Cilium etcd cluster with only three members? Perhaps not worth supporting that edge case, but maybe this restriction should only apply to the two etcd clusters used by apiserver?

Reply:
I think `Forbidden` is worth it just for the details string. I'll switch to that.

Having a different size for the cilium cluster could perhaps be interesting, but since we currently use the internal master DNS entry for this cluster, it won't work (or at least cilium will try to connect to members that don't exist). If someone really wants this, I can look into it later.

Reply:
@johngmyers Which field would you say should be the target here?
There is a similar case when creating an IG with a subnet that doesn't exist. Here we use `NotFound`, but the path is much simpler.
With `Forbidden`, it makes more sense to point to `spec` itself, but not sure that makes that much sense. What do you think?

Reply:
The field is `field.NewPath("objectMeta", "name")`, of the instancegroup. The detail would be something like `instanceGroup "eu-central-1a" must have a member in etcdCluster "main"`.